include/circt/Dialect/FIRRTL/FIRRTLExpressions.td

```diff
@@ -642,3 +642,14 @@ def BitCastOp: FIRRTLOp<"bitcast", [NoSideEffect]> {
 
   let assemblyFormat = "$input attr-dict `:` functional-type($input, $result)";
 }
+
+def PlaceholderOp : FIRRTLOp<"placeholder", [NoSideEffect]> {
```

Comment:
Is this better done as a region op?

Reply:
Or is it some type of "foreign instance"?

Reply:
This is a dummy operation used in `LowerToHW` to allow for use-before-def output here:
https://github.com/llvm/circt/blob/0cbf06173f1e0a2d9cc223f46b890b852e614cc2/lib/Conversion/FIRRTLToHW/LowerToHW.cpp#L2765-L2775

The problem is as follows: to lower a FIRRTL instance input port with an opaque type we cannot generate a temporary wire that later connect ops assign a value to, as we do for non-opaque types. Instead, the value assigned to that port has to be passed to the HW instance directly. `getSingleConnectUserOf` finds this value, but if that value is defined by an op after the instance it hasn't been lowered from FIRRTL to HW yet:
```mlir
%x = firrtl.instance @Foo(in x: !firrtl.opaque<f32>)
%0 = arith.addf ... : f32
%1 = builtin.unrealized_conversion_cast %0 : f32 to !firrtl.opaque<f32>
firrtl.strictconnect %x, %1 : f32
```
To get around this, I create a `PlaceholderOp` as a stand-in for the lowering of the connected value. When doing the lowering for `StrictConnectOp` and the LHS is a placeholder op, the connect _replaces_ the placeholder op with the connected value.

Now that I'm typing this up I realized that instead of a `PlaceholderOp` I might be able to just use an `unrealized_conversion_cast` to directly hook the single connected value up to the HW instance, and then let the cast evaporate once the value is then finally lowered.